peers.rb

```diff
@@ -17,18 +18,148 @@
+  !!@storage.get_id_by_email(email)
+end
+
+def valid_email?(email)
+  return false if email.size == 0
+  (email.split('@').size > 1) && (email.split('.').size > 1)
+end
+
+# ----- LOGIN HELPER METHODS
+
+def valid_credentials?(email, password)
+  @password = @storage.get_password_by_email(email)
+
+  if @password
+    decrypted_password = BCrypt::Password.new(@password)
+    decrypted_password == password
+  else
+    false
+  end
+end
+
+def is_user_login?(id, email)
+  id == @storage.get_id_by_email(email)
+end
+
+# ----- ROUTES
+
 get "/" do
   erb :index
 end
 
-get "/users" do
-  @user = @storage.first_user()
-  return "#{@user["firstname"]} (id: #{@user["pkey"]}) is a user"
+get "/signup" do
+  erb :signup_test, layout: :footer_layout
+end
+
+post "/signup" do
+  email = params[:email]
+  password = params[:password]
+  full_name = params[:full_name]
+
+  if email_in_use?(email)
+    puts "An account with this email already exists, please login"     # Temp. msg to console -> flash msg to be implemented
+    erb :signup_test, layout: :footer_layout
+  elsif !valid_email?(email)
+    puts "Please enter a valid email address"
+    erb :signup_test, layout: :footer_layout
+  elsif password.size == 0
+    puts "Please enter a valid email password"
+    erb :signup_test, layout: :footer_layout
+  elsif full_name.size == 0
+    puts "Please enter a valid full name"
+    erb :signup_test, layout: :footer_layout
+  else
+    password = encrypt_password(password)
+    @storage.add_new_user(email, password, full_name)
+    puts "Your account has been created. Please login using your credentials"
+    redirect "/login"
+  end
+end
+
+get "/login" do
+  erb :login_test, layout: :footer_layout
+end
+
+post "/login" do
+  email = params[:email]
+  password = params[:password]
+
+  if valid_credentials?(email, password)
+    @user = @storage.get_user_by_email(email)
+    puts "Login completed"
+    redirect "/profile/#{@user.id}"
+  else
+    puts "Sorry, your credentials don't exist in our database. Please try again or create an account."
+    erb :login_test, layout: :footer_layout
+  end
+end
+
+get "/logout" do         # To be implemented
+  # So far it just redirects to "/". I tried with '@storage.disconnet' but raises an error
+  puts "Logout complete"
+  redirect "/"
+end
+
+get "/profile/:id" do
+  user_id = params[:id]
+  @user = @storage.get_user_by_id(user_id)
+  if is_user_login?(@user.id, @user.email)               # implemented as:  id == @storage.get_id_by_email(email)
+    erb :profile, layout: :footer_layout
+  else
+    puts "Sorry, you must be login to view this content"
+    erb :login_test, layout: :footer_layout
+  end
+end
+
+get "/profile/:id/edit" do
+  user_id = params[:id]
+  @user = @storage.get_user_by_id(user_id)
+  erb :profile_edit_test, layout: :footer_layout
 end
 
-get "/test/profile/:id" do
+post "/profile/:id/edit" do                # Shall we make this a PUT route instead (see Notions comments on routes section)
```

Comment:
I think HTML form can only make GET and POST requests by default. If we want to use PUT, I think we'd need to override the default behavior on submit, which is very doable using some Javascript code, but perhaps not necessary right now. I say we leave it as is for the time being.